planetiler-core/src/main/java/com/onthegomap/planetiler/Planetiler.java

```diff
@@ -250,6 +253,62 @@ public Planetiler addShapefileSource(String name, Path defaultPath, String defau
     return addShapefileSource(null, name, defaultPath, defaultUrl);
   }
 
+  /**
+   * Adds a new ESRI shapefile directory source that will process all files under {@param basePath} matching
+   * {@param globPattern} using an explicit projection.
+   *
+   * @param projection  the Coordinate Reference System authority code to use, parsed with
+   *                    {@link org.geotools.referencing.CRS#decode(String)}
+   * @param sourceName  string to use in stats and logs to identify this stage
+   * @param basePath    path to the directory containing shapefiles to process
+   * @param globPattern string to match filenames against, as described in {@link FileSystem#getPathMatcher(String)}.
+   * @return this runner instance for chaining
+   * @see ShapefileReader
+   */
+  public Planetiler addShapefileDirectorySource(String projection, String sourceName, Path basePath,
+    String globPattern) {
+    Path dirPath = getPath(sourceName, "shapefile directory", basePath, null);
+    PathMatcher matcher = dirPath.getFileSystem().getPathMatcher("glob:" + globPattern);
+
+    return addStage(sourceName, "Process all files matching " + dirPath + "/" + globPattern,
+      ifSourceUsed(sourceName, () -> {
+        // Since we process many files in parallel, we create a parent timer at the top level.
+        var timer = stats.startStage(sourceName);
+
+        try (var dirWalker = Files.walk(dirPath)) {
+          var pathStream = dirWalker.filter(path -> matcher.matches(path.getFileName()));
+
+          WorkerPipeline.start(sourceName, stats)
+            .readFrom("reader", pathStream::iterator)
+            .addBuffer("read_queue", 1000)
+            .sinkToConsumer("process", config.featureProcessThreads(), path -> {
+              LogUtil.setStage(sourceName, path.getFileName().toString());
+              ShapefileReader.processWithProjection(projection, sourceName, path, featureGroup, config, profile, stats,
+                false);
+              LogUtil.clearStage();
+            })
```

Comment:
Internally `processWithProjection` is kicking off its own worker pipeline.  I think we only want one worker pipeline in order to spread work out appropriately across all cores.  Can we push this handling down into the reader - possibly refactoring `SimpleReader.process` to be multi-file aware? Ideally this will let us support glob patterns for any new kind of reader (geopackage, flatgeobuf, etc.) without having to add special handling for globs for that file type. I'm picturing we get a count from each file ahead of time for progress reporting, then the reader thread iterates through each file and enqueues all of the elements and the rest of the pipeline should remain unaffected.

Reply:
So to make sure I'm following, that would mean that `processWithProjection` takes a `Stream<Path>` or `List<Path>` instead of a single `Path`?

Reply:
Yes exactly, at first I was thinking a `List<Path>` so you could traverse once to get the total count, and a second time to actually extract the elements.  Would that list be too big?